optimizer_spl.py

```diff
@@ -0,0 +1,173 @@
+import numpy as np
+from scipy.interpolate import Akima1DInterpolator
+import matplotlib as mpl
+from matplotlib import pyplot as plt
+np.seterr(divide='ignore', over='raise')
+
+def u(r, m, li):
+    return Akima1DInterpolator(m, li).__call__(r)
+
+def g(r, m, li):
+    return np.exp(-u(r, m, li))
+
+def grad_uu(r, m, li):
+    h = 0.001
+    du = np.ndarray(shape=(len(m),len(r)))
+    
+    for i, j in enumerate(li):
+        li_temp_high = np.copy(li)
+        li_temp_low = np.copy(li)
+        li_temp_high[i] += h
+        li_temp_low[i] -= h
+        
+        u_i_high = u(r, m, li_temp_high)
+        u_i_low = u(r, m, li_temp_low)
+        du[i] = (u_i_high - u_i_low)/(2*h)
+        
+    return du
+
+def grad_ut(r, m, li):
+    h = 0.001
+    du = np.ndarray(shape=(len(m),len(r)))
+    ui_temp_high = np.copy(li)
+    ui_temp_low = np.copy(li)
+    
+    for i, j in enumerate(li):
+        li_temp_high = np.copy(li)
+        li_temp_low = np.copy(li)
+        li_temp_high[i] += h
+        li_temp_low[i] -= h
+        
+        for i, (j, k) in enumerate(zip(li_temp_high, li_temp_low)):
+            ui_temp_high[i] = ui_temp_high[-1] + np.sum(li_temp_high[i:-1])
+            ui_temp_low[i] = ui_temp_low[-1] + np.sum(li_temp_low[i:-1])

```

Comment:
FYI: You could replace this loop with a single command using a cumulative sum (`np.cumsum`), which returns the first value, the sum of the first two values, the sum of the first three values, etc.  You'll need to `np.flip` the array to sum from the end point, and then `np.flip` it back though.  You don't need to actually implement this, just wanted to let you know!